device_backends/include/AsyncDomainsContainer.h

```diff
@@ -18,126 +16,70 @@ namespace ChimeraTK {
   /**
    * The AsyncDomainsContainer has a container with AsyncDomains and is performing actions on all of them.
    *
-   * At the moment, the only common action is sending exceptions. This is implemented via a thread.
+   * The key type to identify the domain is of type size_t, which is the return type of std::hash, so you can easily use
+   * the hashes in case the backend uses a different key type.
+   *
+   * Sending exceptions is implemented via a thread.
    * sendExceptions() is pushing the exception message into a queue and returns immediately.
    * A distributor thread is waiting for data in the queue and is sending the exceptions in all AsyncDomains.
    *
-   * The reason for having a thread is is possible lock order inversion. In the distribution tree, accessor creation
+   * The reason for having a thread is a possible lock order inversion. In the distribution tree, accessor creation
    * must be mutually exclusibe with data distribution, hence locks cannot be avoided. As excaption can occur during
    * data distribution. and backend::setException() is called, this would lead to recursive calls to the distribution
    * tree, which might result in lock order inversions and deadlocks. To avoid this, setException will only put the
    * exception message into the queue and return, allowing the failing distribution call to complete. The exception
    * distribution will then be done by the thread in the AsyncDomainsContainer, after the mutex in the AsyncDomain is
    * free again.
    */
-  template<typename KeyType>
-  class AsyncDomainsContainer : public AsyncDomainsContainerBase {
+  class AsyncDomainsContainer {
    public:
-    explicit AsyncDomainsContainer();
-    ~AsyncDomainsContainer() override;
+    ~AsyncDomainsContainer();
+
+    /** Request the sending of exceptions. This function stores the request and returns immediately.
+     *  The actual exception distribution is done asynchronously and has not necessarily finished when the function call
+     * returns. Use isSendingExceptions() to check whether the exception distribution has finished.
+     */
+    void sendExceptions(const std::string& exceptionMessage);
 
-    void sendExceptions(const std::string& exceptionMessage) override;
+    /** Check whether an exception distribution is started and not completed yet. */
+    bool isSendingExceptions() { return _isSendingExceptions; }
 
     /**
      * Adds an AyncDomain to the container. If there is already an entry for that key, and that weak pointer can be
      * locked, it throws a ChimeraTK::logic_error. This function must only be called if the either an entry for that key
      * does not exist, or the according share pointer is gone.
      */
-    void addAsyncDomain(KeyType key, const boost::shared_ptr<AsyncDomain>& asyncDomain);
+    void addAsyncDomain(size_t key, const boost::shared_ptr<AsyncDomain>& asyncDomain);
```

Comment:
Maybe to clarify in the description: The `asyncDomain` must be empty when adding, so there must not yet be any subscribers attached to it (otherwise there might be race conditions with exceptions).

I am wondering whether it is possible to write some kind of "getOrCreateAsyncDomain" so we don't need handle these details correctly in the calling code...

Reply:
The AsyncDomain does not have to be empty because it has a lock itself. If an accessor is already in the AsyncDomain, it will see the exception if the domain was active. If the exception comes first, the domain is inactive and the new accessor does not see an exception. Distributing the exception, creating the accessor and turning on/add the active flag is all happening under the lock inside the AsyncDomain.

However, there is a race condition here: after getAsyncDomain() returns nullptr, a different thread might have added the domain already, so addAsyncDomain() will throw. This violates the specification because it says that you can always check if logic_error will be thrown. However, this is also true for throwing a logic error when a device is closed and you try to read an accessor. Another thread might have closed it after you have just checked.

Reply:
> However, this is also true for throwing a logic error when a device is closed and you try to read an accessor. Another thread might have closed it after you have just checked.

You keep repeating this argument for making anything else also not thread-safe. We agreed on `close()` not being thread safe, but everything else must be thread safe.

Anyway: It is a runtime_error, not a logic_error, probably neither is correct. A runtime_error means, the remote device (or the connection to it) does not behave as expected, which is clearly not the case. A logic_error means, the application programmer has done something wrong (which implies that he could have known this beforehand) - in that case at least the message would not be helpful at all, since the application programmer does not even know that there is an AsyncDomain involved.

This needs to be solved differently.

Reply:
The function has been replaced with  ```getOrCreateAsyncDomain()```, which creates the async domain under the container lock and removes the race condition.